source/Timer.h

```diff
@@ -0,0 +1,126 @@
+/* Timer.h
+Copyright (c) 2023 by Timothy Collett
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef TIMER_H_
+#define TIMER_H_
+
+#include "MissionAction.h"
+#include "System.h"
+
+#include <cstdint>
+#include <map>
+#include <string>
+
+class DataNode;
+class DataWriter;
+class Planet;
+class PlayerInfo;
+class UI;
+
+
+
+// Class representing a timer for triggering mission actions
+// This requires a specification of how long to wait (base + randint(rand)),
+// a system to wait in, and whether to reset the timer upon leaving the system
+// and can optionally require the player to be idle (no acceleration), and/or
+// within or beyond a certain proximity to the system center or a
+// specific (named) system object.
+class Timer {
+public:
+	Timer() = default;
+	Timer(const DataNode &node, const Mission *mission);
+	// Set up the timer from its data file node.
+	void Load(const DataNode &node, const Mission *mission);
+	// Note: the Save() function can assume this is an instantiated Timer, not a template,
+	// so the time to wait will be saved fully calculated, and with any elapsed time subtracted.
+	void Save(DataWriter &out) const;
+
+	// Get whether the timer is optional to complete.
+	bool IsOptional() const;
+	// Get whether the timer is currently active.
+	bool IsComplete() const;
+
+	// Calculate the total time to wait, including any random value.
+	Timer Instantiate(std::map<std::string, std::string> &subs,
+					const System *origin, int jumps, int64_t payload) const;
+	// Progress the timer within the main loop.
+	void Step(PlayerInfo &player, UI *ui, Mission *mission);
```

Comment:
I'd say just pass the mission as a reference. I know that MissionAction accepts mission as a pointer, but that's because GameAction (a member of MissionAction) can potentially be given a nullptr for mission in other locations. In the case of Timer, we know we're always getting something. A Timer can't exist except as a part of a mission, after all.

Could make it const too since Timer only needs the mission so it can pass it off to MissionAction, and MissionAction takes a `const Mission *`.
```suggestion
	void Step(PlayerInfo &player, UI *ui, const Mission &mission);
```